File: backend/common/src/main/java/io/quizchef/common/api/ApiError.java
package io.quizchef.common.api;

import io.quizchef.common.correlation.CorrelationKeys;
import java.time.Instant;
import java.util.List;
import org.slf4j.MDC;

/**
 * The single error response format for every QuizChef API.
 *
 * @param code          stable machine-readable error code
 * @param message       human-readable summary, safe to display
 * @param timestamp     when the error occurred (UTC)
 * @param correlationId the request's correlation id, so a support report or a
 *                      fatal-error dialog can be matched back to server logs
 *                      (Phase 3 PR #2); {@code null} only when an error is
 *                      built outside a request thread
 * @param fieldErrors   per-field validation failures, empty unless the
 *                      request failed bean validation
 */
public record ApiError(
        String code,
        String message,
        Instant timestamp,
        String correlationId,
        List<ApiFieldError> fieldErrors
) {

    public ApiError {
        fieldErrors = List.copyOf(fieldErrors);
    }

    public static ApiError of(String code, String message) {
        return new ApiError(code, message, Instant.now(), currentCorrelationId(), List.of());
    }

    public static ApiError validation(String message, List<ApiFieldError> fieldErrors) {
        return new ApiError("validation.failed", message, Instant.now(), currentCorrelationId(), fieldErrors);
    }

    private static String currentCorrelationId() {
        return MDC.get(CorrelationKeys.CORRELATION_ID_MDC_KEY);
    }
}

File: frontend/src/api/apiError.ts
import type { ApiFieldError } from "@/types/api";

/**
 * The one error shape the rest of the frontend sees. Every axios failure is
 * mapped to this in the response interceptor, so callers switch on the
 * backend's stable error `code` (RFC-002/003/004) — never on transport
 * details.
 */
export class ApiClientError extends Error {
  readonly code: string;
  /** HTTP status, or null when the request never reached the server. */
  readonly status: number | null;
  readonly fieldErrors: ApiFieldError[];
  /**
   * The backend's request correlation id (Phase 3 PR #2), so a fatal-error
   * dialog can show a reference a support report can match back to server
   * logs. Null when the request never reached the server.
   */
  readonly correlationId: string | null;

  constructor(
    code: string,
    message: string,
    status: number | null,
    fieldErrors: ApiFieldError[] = [],
    correlationId: string | null = null
  ) {
    super(message);
    this.name = "ApiClientError";
    this.code = code;
    this.status = status;
    this.fieldErrors = fieldErrors;
    this.correlationId = correlationId;
  }
}

export function isApiClientError(error: unknown): error is ApiClientError {
  return error instanceof ApiClientError;
}

/** A human-readable message for any error, with a safe fallback. */
export function errorMessage(error: unknown): string {
  if (isApiClientError(error)) {
    return error.message;
  }
  if (error instanceof Error && error.message) {
    return error.message;
  }
  return "Something went wrong. Please try again.";
}
